extensions/amp-ad-network-cloudflare-impl/0.1/amp-ad-network-cloudflare-impl.js

```diff
@@ -62,18 +54,64 @@ export class AmpAdNetworkCloudflareImpl extends AmpA4A {
     return ['cloudflare'];
   }
 
+  /**
+   * Handle variable replacements
+   *
+   * @param {string} str input string to process
+   * @param {?Object<string, string>} values to use in the replacements
+   * @return {string} result with replaced tokens
+   */
+  replacements(str, values) {
+    if (!str) {
+      return str;
+    }
+
+    // allow injection of width and height as parameters
+    str = str.replace(/SLOT_WIDTH/g, values.slotWidth);
+    str = str.replace(/SLOT_HEIGHT/g, values.slotHeight);
+
+    return str;
+  }
+
   /** @override */
   getAdUrl() {
-    // You can optionally use data from this context to augment the URL,
-    // to pass data back to the network, to aid in ad selection.  For example:
-    //   const rect = this.getIntersectionElementLayoutBox();
-    // can extract the layout size and be used to pass layout hints to the server.
+    const rect = this.getIntersectionElementLayoutBox();
+    const el = this.element;
+
+    const network = el.getAttribute('data-cf-network');
+    const a4a = el.getAttribute('data-cf-a4a') !== 'false';
+    const base = NETWORKS[network].base;
+
+    // generate URL for ad creative
+    let src = el.getAttribute('src');
+    if (src == '' || src[0] != '/') {
+      // ensure that we start from the root
+      src = '/' + src;
+    }
+    let url = base + (a4a ? '/_a4a' : '') + src;
+
+    // compute replacement values
+    const values = {
+      slotWidth: (rect.width || 0).toString(),
+      slotHeight: (rect.height || 0).toString(),
+    };
+
+    // encode for safety
+    url = encodeURI(this.replacements(url, values));
```

Comment:
Since the only things that aren't controlled in the code base are `slotWidth` and `slotHeight`, should we just `encodeURIComponent` those?